README.md
# TSSr: an R/Bioconductor package for comprehensive analyses of transcription start sit (TSS) data

##### Zhaolian Lu, Keenan Berry, Zhenbin Hu, Yu Zhan, Tae-Hyuk Ahn, Zhenguo Lin


## 1. Intruduction

Documentation is also available on GitHub Pages: https://github.com/Linlab-slu/TSSr

 TSSr is an R package that provides rich functions for mapping transcription start sites (TSSs) and characterizations of structures and activities of core promoters based on all types of TSS sequencing data, such as cap analysis of gene expression (CAGE) (Takahashi, Lassmann et al. 2012), no-amplification non-tagging CAGE libraries for Illumina next-generation sequencers (nAnT-iCAGE) (Murata, Nishiyori-Sueki et al. 2014), a Super-Low Input Carrier-CAGE (SLIC-CAGE) (Cvetesic, Leitch et al. 2018), NanoCAGE (Cumbie, Ivanchenko et al. 2015), TSS-seq (Malabat, Feuerbach et al. 2015), transcript isoform sequencing (TIF-seq) (Pelechano, Wei et al. 2013), transcript-leaders sequencing (TL-seq) (Arribere and Gilbert 2013), precision nuclear run-on sequencing (PRO-Cap) (Mahat, Kwak et al. 2016), and GRO-Cap/5’GRO-seq (Kruesi, Core et al. 2013).
 
 TSSr package provides a comprehensive workflow on TSS data starts from accurate identification of TSS locations, clustering TSSs within small genomic regions to infer core promoters, and quantifications of transcriptional activity, as well as specialized downstream analyses including quantification of core promoter shape, asigning TSS cluster to genes, gene differential expression, core promoter shift. TSSr can take multiple formats of files as input, such as Binary Sequence Alignment Mao (BAM) files (single-ended or paired-ended), Browser Extension Data (bed) files, BigWig files, ctss files or tss tables. TSSr generates various types of TSS or core promoter track files which can be visualized in the UCSC Genome Browser or Integrative Genomics Viewer (IGV). TSSr also exports various raw and processed result tables and visualization plots. Multiple cores are supported on Linux or Mac platforms.
 
## 2. Pre-requisites:

* R version
  * Download R (>3.4.0) version from CRAN.
    * Windows: https://cran.r-project.org/bin/windows/base/
    * Mac OS X: https://cran.r-project.org/bin/macosx/
    * Linux: https://cran.r-project.org/bin/linux/

- Rsamtools package

  * install by using the following R commands:

        if (!requireNamespace("BiocManager", quietly = TRUE))
        install.packages("BiocManager")
        BiocManager::install("Rsamtools")


- GenomicRanges

  * install by using the following R commands:

        if (!requireNamespace("BiocManager", quietly = TRUE))
        install.packages("BiocManager")
        BiocManager::install("GenomicRanges")

- GenomicFeature package

  * install by using the following R commands:

        if (!requireNamespace("BiocManager", quietly = TRUE))
        install.packages("BiocManager")
        BiocManager::install("GenomicFeatures")

- Gviz package

  * install by using the following R commands:

        if (!requireNamespace("BiocManager", quietly = TRUE))
        install.packages("BiocManager")
        BiocManager::install("Gviz")  

- rtracklayer package

  * install by using the following R commands:

        if (!requireNamespace("BiocManager", quietly = TRUE))
        install.packages("BiocManager")
        BiocManager::install("rtracklayer")

- DESeq2 package:

  * install by using the following R command:

        if (!requireNamespace("BiocManager", quietly = TRUE))
        install.packages("BiocManager")
        BiocManager::install("DESeq2")

- BSgenome package

  * install by using the following R command:

        if (!requireNamespace("BiocManager", quietly = TRUE))
        install.packages("BiocManager")
        BiocManager::install("BSgenome")

- data.table package

  * install by using the following R command:

        if (!requireNamespace("data.table", quietly = TRUE))
        install.packages("data.table")

- stringr package

  * install by using the following R command:

        if (!requireNamespace("stringr", quietly = TRUE))
        install.packages("stringr")
	

## 3. Installing TSSr Package:

To install the TSSr package all the prerequisites above need to be installed.
After confirming those packages are installed, you can install the development version directly from GitHub using devtools:

        devtools::install_github("Linlab-slu/TSSr")

## 4. Input data for TSSr

* Input data files

  TSSr accepts two types of input data: read alignment files or TSS tables. 
  The read alignment files in compressed binary alignment map (BAM) format are required if users intend to call TSSs from mapped sequencing data. BAM files can be derived from mapping of either paired-end or single-end TSS sequencing reads. 
  TSSr also accepts TSS tables generated by TSSr or other bioinformatics tools as input data. A TSS table can be a tab-delimited text file, a BigWig (bw) binary type file, or browser extensible data (bed) type file. A tab-delimited TSS table contains chromosome ID, genomic coordinates, strand information, and raw or normalized read counts of each sample. An example of TSS table can be downlaoded from http://zlinlab.org/TSSr/ALL.samples.TSS.raw.txt. 
  

* Reference genome

  The reference genome stored as a BSgenome data package must be provided to determine the presence of genome uncoded G cap in sequencing read during TSS calling. For available BSgenome genome objects, please check (https://bioconductor.org/packages/BSgenome). In case the BSgenome object is unavailable for a species in BSgenome package (not in the list returned by BSgenome::available.genomes() function), a custom genome has to be built and installed before run through this package. (See the vignette “How to forge a BSgenome data package” within the BSgenome package for instructions). A genome annotation file (GTF or GFF file) is required to assign TSS clusters to genes or genomic features by using the annotateCluster function. The chromosome names in BSgenome, genome annotation file as well as the reference genome used for mapping sequencing reads must be matched.  

* S4 object

  TSSr uses S4 object to store all input files and arguments and generate downstream analysis results.

## 5. Usage:

* Launch TSSr

        library(TSSr)
      
* Users may explore the fucntions of TSSr using provided example TSSr object "exampleTSSr". Skip this step if the user plan to generate a new TSSr object from bam or TSS table files 

        myTSSr <- exampleTSSr 
        
* Generating a new TSSr object from BAM or TSS table files. The first step is to provide the path and file names of input files "inputFiles". Four example BAM files (S01.sorted.bam, S02.sorted.bam, S03.sorted.bam, S04.sorted.bam) can be downloaded from http://www.zlinlab.org/TSSr.html. These BAM files were obtained from a budding yeast Saccharomyces cerevisiae based on CAGE sequencing (Lu and Lin 2019). The CAGE reads were mapped to the reference genome of S. cerevisiae (R64-2-1) using HISAT2 (Kim, Langmead et al. 2015). To reduced sample file size, each BAM file only includes sequencing reads mapped to two chromosomes (Chr I and Chr II). A TSS table genereated from the four BAM files can be downloaded from http://zlinlab.org/TSSr/ALL.samples.TSS.raw.txt. 

        inputFiles <- c("S01.sorted.bam", "S02.sorted.bam", "S03.sorted.bam", "S04.sorted.bam") # if use a TSS table: inputFiles <- "ALL.samples.TSS.raw.txt"
        inputFilesType <- "bam" # set “inputFilesType” as “bamPairedEnd” for paired-end BAM files, and as "TSStable" if input data is a TSS table 
	
* Install required BSgenome object. The example CAGE data was obtained from Saccharomyces cerevisiae. The BSgenome object for S. cerevisiae “BSgenome.Scerevisiae.UCSC.sacCer3” is available from BSgenome Bioconductor package (https://bioconductor.org/packages/BSgenome). 

        if (!requireNamespace("BSgenome.Scerevisiae.UCSC.sacCer3", quietly = TRUE))
        BiocManager::install("BSgenome.Scerevisiae.UCSC.sacCer3")
        library(BSgenome.Scerevisiae.UCSC.sacCer3) 
	
* Provide path and file name of genome annotation file. The example genome annotation of S. cerevisiae (R64-2-1) was originally obtained from the Saccharomyces Genome Database. It ban be downlaoded from http://zlinlab.org/TSSr/saccharomyces_cerevisiae_R64-2-1.gff .

        gffFile <- "saccharomyces_cerevisiae_R64-2-1.gff" 
	
* Creating a new TSSr object using the constructer "new" function. In addition to the input data file and type, BS genome object name, and genome annotaion file, this step also requires users to provide more information about the samples, including "sampleLabels", "sampleLabelsMerged", and "mergeIndex" and "organismName". In the example data, sample lables "SL01, SL02, SL03 and SL04" were used for "S01.sorted.bam, S02.sorted.bam, S03.sorted.bam, S04.sorted.bam", respectively (sampleLabels = c("SL01","SL02","SL03","SL04")). If a TSS table is used as input data, the sample lable of each sample is its column name in the TSS table. "sampleLabelsMerged" will be used if the user need to merge TSS signals from biological replicates into a single dataset. In this case, SL01 and SL02 are two biological replicates obtained from S. cerevisiae cells grown in rich medium (control), while SL03 and SL04 were obtained by treating S. cerevisiae with α factor (treat). Thus, sampleLabelsMerged = c("control","treat") and mergeIndex = c(1,1,2,2), which means SL01 and SL02 will be merged as group 1 "control", and SL03 and SL04 will be merged as group 2 "treat". "organismName" is the species name of the samples, which will be used to annotate TSS clusters by "annotateCluster".

        myTSSr <- new("TSSr", genomeName = "BSgenome.Scerevisiae.UCSC.sacCer3"
	              ,inputFiles = inputFiles
	              ,inputFilesType = inputFilesType
	              ,sampleLabels = c("SL01","SL02","SL03","SL04")
	              ,sampleLabelsMerged = c("control","treat")
	              ,mergeIndex = c(1,1,2,2)
	              ,refSource = gffFile
	              ,organismName = "saccharomyces cerevisiae")
	
	To display the available slots in the created TSSr object by typing the TSSr object name. Most slots in the newly created TSSr object are empty at this point since no analysis has been conducted.:
	
	     myTSSr
	      
	      # An object of class "TSSr"
        # Slot "genomeName":
        #   [1] "BSgenome.Scerevisiae.UCSC.sacCer3"
        # 
        # Slot "inputFiles":
        #   [1] "S01.sorted.bam" "S02.sorted.bam" "S03.sorted.bam" "S04.sorted.bam"
        # 
        # Slot "inputFilesType":
        #   [1] "bam"
        # 
        # Slot "sampleLabels":
        #   [1] "SL01" "SL02" "SL03" "SL04"
        # 
        # Slot "sampleLabelsMerged":
        #   [1] "control" "treat"  
        # 
        # Slot "librarySizes":
        #   integer(0)
        # 
        # Slot "TSSrawMatrix":
        #   data frame with 0 columns and 0 rows
        # 
        # Slot "mergeIndex":
        #   numeric(0)
        # 
        # Slot "TSSmergedMatrix":
        #   data frame with 0 columns and 0 rows
        # 
        # Slot "TSSnormalizedMatrix":
        #   data frame with 0 columns and 0 rows
        # 
        # Slot "TSSfilteredMatrix":
        #   data frame with 0 columns and 0 rows
        # 
        # Slot "tagClusters":
        #   list()
        # 
        # Slot "consensusClusters":
        #   list()
        # 
        # Slot "clusterShape":
        #   list()
        # 
        # Slot "refSource":
        #   [1] "saccharomyces_cerevisiae.SGD.gff"
        # 
        # Slot "organismName":
        #   [1] "saccharomyces cerevisiae"
        # 
        # Slot "assignedClusters":
        #   list()
        # 
        # Slot "unassignedClusters":
        #   list()
        # 
        # Slot "DEtables":
        #   list()
        # 
        # Slot "PromoterShift":
        #   list()
	
	
	
* TSS calling from bam files or retrieving TSS data from TSS table using "getTSS" function. The "getTSS" function retrieves TSS coordinates and tag counts from multiple samples into one data table. Before TSS calling, TSSr removes reads that are below certain sequencing quality and mapping quality. The default threshold for Phred quality score is 10, and mapping quality (MAPQ score) is 20. Users may change these parameters by setting different values for “sequencingQualityThreshold” and “mappingQualityThreshold” when running the “getTSS” function.  
	
        getTSS(myTSSr)
	
	TSS calling from bam files or retrieving TSS data from TSS table         
        myTSSr@TSSrawMatrix
        
        # chr    pos strand SL01 SL02 SL03 SL04
        # 1:  chrI   1561      +    0    0    0    1
        # 2:  chrI   1823      +    1    0    0    0
        # 3:  chrI   1828      +    0    0    0    2
        # 4:  chrI   1830      +    0    0    1    0
        # 5:  chrI   1831      +    0    0    0    1
        # ---                                        
        # 171537: chrII 812483      -    0    0    0    1
        # 171538: chrII 812789      -    0    0    0    2
        # 171539: chrII 812802      -    0    0    0    2
        # 171540: chrII 812805      -    0    0    0    1
        # 171541: chrII 812818      -    0    0    0    1

* TSSrawMatrix contains raw read counts of each called TSS. Raw TSS count can be exported to TSS tables can be used as inout for future analysis using TSSr. The output TSS table is saved as "ALL.samples.TSS.raw.txt" in the working directory.    

        exportTSStable(myTSSr, data = "raw", merged = "FALSE")
	
* The "plotCorrelation" function is used to calculate the pairwise correlation coefficients and plot pairwise scatter plots of TSS counts. A subset of samples can also be specified to display the pairwise correlations. Three correlation methods are supported: “pearson”, “kendall”, or “spearman”.
	
        plotCorrelation(myTSSr, samples = "all")
        
![01_TSS_correlation_plot_of_all_samples](https://github.com/Linlab-slu/TSSr/raw/master/vignettes/figures/01_TSS_correlation_plot_of_all_samples.png)

 * To further explore the variations present in the TSS dataset and identify which samples are similar to each other and which samples are very different, we can apply plotPCA function to plot principle component analysis among all samples. plotPCA will make a biplot which visualizes both how samples relate to each other in terms of PC1 and PC2 and simultaneously show how each variable contributes to each principal component.
  
        plotTssPCA(myTSSr, TSS.threshold=10)

![02_PCA_plot](https://raw.githubusercontent.com/Linlab-slu/TSSr/master/vignettes/figures/02_PCA_plot.png)

* Based on the calculated correlations, two replicates are highly correlated. To facilitate the downstream analysis and comparisons between different growth conditions, we can merge the two biological replicates for each growth condition together with mergeSamples function. mergeIndex argument directs which samples will be merged and how the final dataset will be ordered accordingly.
  
        mergeSamples(myTSSr)
        
        myTSSr@TSSprocessedMatrix
        
        # chr    pos strand control treat
        # 1:  chrI   1561      +       0     1
        # 2:  chrI   1823      +       1     0
        # 3:  chrI   1828      +       0     2
        # 4:  chrI   1830      +       0     1
        # 5:  chrI   1831      +       0     1
        # ---                                  
        # 171537: chrII 812483      -       0     1
        # 171538: chrII 812789      -       0     2
        # 171539: chrII 812802      -       0     2
        # 171540: chrII 812805      -       0     1
        # 171541: chrII 812818      -       0     1

  To return library sizes (number of total sequenced tags) of each sample in TSSr object in the specified order (note: order is specified in mergeSample function):

        myTSSr@librarySizes
        # [1] 3353873 5563702

  Library sizes among different samples are different. To make samples comparable, we use normalizeTSS function to scale TSS raw counts by tags per million (TPM).

        normalizeTSS(myTSSr)
        
  There is a great amount of TSSs with low weak transcriptional signals. To filter out low-fidelity TSSs, we use filterTSS function to remove TSSs below the specified threshold. Two filtering methods are supported: “poisson” or “TPM”. 

        filterTSS(myTSSr, method = "TPM",tpmLow=0.1)
    
        myTSSr@TSSprocessedMatrix
        
        # chr    pos strand  control    treat
        # 1:  chrI   1561      + 0.000000 0.179736
        # 2:  chrI   1823      + 0.298163 0.000000
        # 3:  chrI   1828      + 0.000000 0.359473
        # 4:  chrI   1830      + 0.000000 0.179736
        # 5:  chrI   1831      + 0.000000 0.179736
        # ---                                      
        # 171537: chrII 812483      - 0.000000 0.179736
        # 171538: chrII 812789      - 0.000000 0.359473
        # 171539: chrII 812802      - 0.000000 0.359473
        # 171540: chrII 812805      - 0.000000 0.179736
        # 171541: chrII 812818      - 0.000000 0.179736

  The processed TSS matrix can be exported to either text tables or bedGraph/BigWig tracks which can be visualized in the UCSC Genome Browser or Integrative Genomics Viewer (IGV).    
        exportTSStoBedgraph(myTSSr, data = "processed", format = "bedGraph")
        exportTSStoBedgraph(myTSSr, data = "processed", format = "BigWig")
        
* TSS clustering

  TSSs within a small genomic region are likely regulated by the sample regulatory elements, and thus can be clustered together as tag clusters (TCs). The simplest and widely used algorithm is distance-based clustering (disclu) in which neighboring TSSs are grouped together if they are closer than a specified distance (Haberle, Forrest et al. 2015). Since disclu algorithm ignores TSS signal distribution within one cluster, it might potentially generate extra big clusters with more than one peaks if the distance threshold is too big. However, reducing distance threshold might create many small clusters. Thus, we developed a peak-based clustering algorithm in which TSSs are clustered based on strong TSS signals which are identified as peaks.
We will perform peak-based clustering with clusterTSS function using 100bp as the minimal range within which there is at most one peaks and use an extension distance 30bp to join neighboring TSSs into the peak defined clusters. We implement another layer of local filtering which rules out weak TSS signals downstream of peaks that are potentially brought out from recapping and are always considered as noise. After clustering, clusters below cluster Threshold will be filtered out. clusterTSS function generates a set of clusters for each sample separately. In each cluster, clusterTSS function returns genomic coordinates, sum of TSS tags, dominate TSS coordinate, a lower (q0.1) and an upper (q0.9) quantile coordinates, and interquantile widths.
This clustering step might be slow especially when the number of TSSs is in millions. Using multicores is highly recommended.

        clusterTSS(myTSSr, method = "peakclu",peakDistance=100,extensionDistance=30
	           ,localThreshold = 0.02,clusterThreshold = 1
	           ,useMultiCore=FALSE, numCores=NULL)
	           
        myTSSr@tagClusters
        
        # $`control`
        # cluster  chr start  end strand dominant_tss    tags tags.dominant_tss q_0.1 q_0.9 interquantile_width
        # 1:    1 chrI  6530  6564   +     6548  8.944883     3.577953  6548  6562         15
        # 2:    2 chrI  7166  7189   +     7169  1.192652     0.596326  7166  7189         24
        # 3:    3 chrI  8084  8087   +     8084  1.192652     0.596326  8084  8087          4
        # 4:    4 chrI  9325  9411   +     9327  4.770607     1.192651  9327  9391         65
        # 5:    5 chrI  9442  9479   +     9442  3.279792     1.788977  9442  9468         27
        # ---                                                        
        # 3485:  3485 chrII 810641 810721   -    810641 77.224157     21.765881 810641 810719         79
        # 3486:  3486 chrII 810760 811032   -    810916 46.513400     5.963255 810796 810962         167
        # 3487:  3487 chrII 811067 811170   -    811112 11.330187     8.348557 811111 811131         21
        # 3488:  3488 chrII 811200 811384   -    811291 29.518121     6.559581 811239 811366         128
        # 3489:  3489 chrII 811446 811512   -    811486 607.059367    313.965377 811485 811494         10
        # 
        # $treat
        # cluster  chr start  end strand dominant_tss    tags tags.dominant_tss q_0.1 q_0.9 interquantile_width
        # 1:    1 chrI  1828  1841   +     1828  1.078417     0.359473  1828  1841         14
        # 2:    2 chrI  6521  6564   +     6559  4.673145     1.617628  6530  6559         30
        # 3:    3 chrI  7096  7118   +     7100  1.797363     0.539209  7096  7114         19
        # 4:    4 chrI  8061  8087   +     8061  1.797363     0.539209  8061  8087         27
        # 5:    5 chrI  9327  9476   +     9359 10.604449     1.617628  9327  9452         126
        # ---                                                        
        # 3382:  3382 chrII 809707 809710   -    809707  1.258155     0.898682 809707 809710          4
        # 3383:  3383 chrII 810641 810736   -    810641 121.501831     29.297040 810641 810719         79
        # 3384:  3384 chrII 810773 811031   -    810916 67.401151     14.738388 810794 810963         170
        # 3385:  3385 chrII 811108 811377   -    811112 69.557986     14.019442 811112 811366         255
        # # 3386:  3386 chrII 811472 811503   -    811486 449.700575    234.196583 811486 811494          9

		    exportClustersTable(myTSSr, data = "assigned")
		    
		    exportClustersToBed(myTSSr, data = "tagClusters")

* Aggregating consensus clusters

  TSSs are clustered into tag clusters for each sample individually and they are often sample-specific. To make clusters comparable between samples, we will use consensusCluster function to generate a single set of consensus clusters. Similarly to clusterTSS function, consensusCluster function also returns genomic coordinates, sum of TSS tags, dominate TSS coordinate, a lower (q0.1) and an upper (q0.9) quantile coordinates, and interquantile widths for each consensus cluster in each sample.
  
        consensusCluster(myTSSr, dis = 50, useMultiCore = FALSE)
		    
  Tag clusters and consensus cluster with quantile positions can be exported to either text tables or BED tracks which can be visualized in the UCSC Genome Browser and IGV. 

		    exportClustersToBed(myTSSr, data = "consensusClusters")


* Core promoter shape

  According to the distribution of TSSs within a core promoter (cluster), termed as core promoter shape, core promoters were generally classified into sharp core promoters and broad core promoters. TSSr implements three methods to characterize core promoter shape. The simplest way is to use interquantile width representing core promoter shape. plotInterQuantile function plots interquantile width of each sample.

        plotInterQuantile(myTSSr,samples = "all",tagsThreshold = 1)
    
![03_Interquantile_plot_of_ALL_samples](https://github.com/Linlab-slu/TSSr/raw/master/vignettes/figures/03_Interquantile_plot_of_ALL_samples.png){width=50%}

  Another way to characterize core promoter shape is shape index (SI) which is determined by the probabilities of tags at every TSSs within one cluster (Hoskins, Landolin et al. 2011). SI is calculated using shapeCluster function with method set as “SI”. The greater value represents the sharper core promoter. SI is 2 representing singletons. Genome-wide SI score can be plotted with plotShape function.

  By integrating both inter quantile width and the observed probabilities of tags at every TSSs within a cluster, we developed a new metric called promoter shape score (PSS) to describe core promoter shape (Lu and Lin 2019). PSS can be calculated using using shapeCluster function with method set as “PSS”. The smaller value represents the sharper core promoter. PSS is 0 representing singletons. Genome-wide PSS score can be plotted with plotShape function.
  
        shapeCluster(myTSSr,clusters = "consensusClusters", method = "PSS",useMultiCore= FALSE, numCores = NULL)

        myTSSr@clusterShape
        
        # $`control`
        # cluster chr start end strand dominant_tss  tags tags.dominant_tss q_0.1 q_0.9 interquantile_width shape.score
        # 1:  432 chrII 127 175  +   150 3.279792   0.894488 145 171     27 -0.41938195
        # 2:  433 chrII 4016 4045  +   4016 2.087141   0.596326 4016 4045     30 -0.23592635
        # 3:  434 chrII 4299 4318  +   4300 1.192652   0.596326 4299 4318     20 0.50000000
        # 4:  435 chrII 5328 5359  +   5328 1.788977   1.192651 5328 5359     32 0.74837083
        # 5:  436 chrII 5577 5577  +   5577 1.788977   1.788977 5577 5577     1 2.00000000
        # ---                                 
        # 3484: 2526 chrI 222756 223117  -  222884 30.114450   2.385302 222847 223076     230 -2.83272807
        # 3485: 2528 chrI 223374 223416  -  223395 2.981629   1.490814 223374 223416     43 0.03903595
        # 3486: 2529 chrI 223656 223696  -  223660 1.490815   0.596326 223656 223696     41 0.07807191
        # 3487: 2530 chrI 225020 225056  -  225056 1.192652   0.596326 225020 225056     37 0.50000000
        # 3488: 2532 chrI 227236 227274  -  227274 2.385303   0.894488 227236 227274     39 -0.15563906
        # 
        # $treat
        # cluster chr start end strand dominant_tss  tags tags.dominant_tss q_0.1 q_0.9 interquantile_width shape.score
        # 1:  436 chrII 5525 5587  +   5577 2.875781   1.977101 5534 5577     44 0.9107699
        # 2:  437 chrII 5895 5955  +   5955 1.797364   0.718946 5920 5955     36 0.1634083
        # 3:  438 chrII 7626 7682  +   7649 7.189456   1.258155 7630 7658     29 -0.8540706
        # 4:  439 chrII 8410 8455  +   8425 4.313673   1.437892 8425 8452     28 -0.4516085
        # 5:  440 chrII 8492 8584  +   8550 12.401807   2.875783 8513 8557     45 -1.0242498
        # ---                                 
        # 3380: 2526 chrI 222797 222969  -  222851 17.614164   4.493411 222851 222961     111 -1.4326016
        # 3381: 2527 chrI 223326 223327  -  223326 1.258155   0.898682 223326 223327     2 1.1368794
        # 3382: 2529 chrI 223659 223697  -  223681 2.156836   0.718946 223659 223696     38 -0.3685225
        # 3383: 2531 chrI 226791 226810  -  226805 1.437890   0.359473 226791 226810     20 -0.5000000
        # 3384: 2532 chrI 227249 227300  -  227270 6.290774   1.977101 227258 227276     19 -0.2658030

        plotShape(myTSSr)
    
![04_Shape_plot_of_ALL_samples](https://github.com/Linlab-slu/TSSr/raw/master/vignettes/figures/04_Shape_plot_of_ALL_samples.png){width=50%}

		    exportShapeTable(myTSSr)

* Annotation

  TSSr identifies TSSs and therefore detect core promoters. However, the current databases are mainly on gene levels instead of TSS or core promoter levels. annorateCluster function is used to associate core promoters to genes. Since annotated genes in Saccharomyces cerevisiae GFF annotation file don’t contain 5’ untranslated region (UTR), that is, the first genomic position of each gene in annotations is start position of translation. Therefore, in order to associate clusters to genes, we specify the upstream distance as 1000 and downstream distance as 0. The 1000 bp region is defined as core promoter region. In case of overlapping with upstream genes, we specify upstreamOverlap argument as 500 representing only considering the first 500 bp regions as core promoter region if it is overlapped with upstream genes.
To reduce transcriptional or technical noise of small clusters downstream a strong cluster, we apply filterCluster argument and set filterClusterThreshold as 0.02, representing clusters in which total tags lower than the strong cluster*0.02 will be filtered out.

        annotateCluster(myTSSr,clusters = "consensusClusters",filterCluster = TRUE,
	                filterClusterThreshold = 0.02, annotationType = "genes"
	                ,upstream=1000, upstreamOverlap = 500, downstream = 0)
	                
        myTSSr@assignedClusters
        
        # $`control`
        # cluster chr start end strand dominant_tss  tags tags.dominant_tss q_0.1 q_0.9 interquantile_width  gene
        # 1:  6 chrI 9325 9411  +   9327 4.770607   1.192651 9327 9391     65 YAL066W
        # 2:  7 chrI 9442 9479  +   9442 3.279792   1.788977 9442 9468     27 YAL066W
        # 3:  8 chrI 11253 11343  +  11329 39.059326   16.995277 11275 11329     55 YAL064W-B
        # 4:  44 chrI 31027 31151  +  31108 46.215232   16.398951 31108 31145     38 YAL062W
        # 5:  45 chrI 31187 31263  +  31242 307.405801  117.774287 31212 31242     31 YAL062W
        # ---                                 
        # 864: 4493 chrII 804562 804593  -  804591 1.490814   0.894488 804562 804593     32 YBR298C
        # 865: 4494 chrII 804855 804975  -  804895 250.456717   68.875595 804868 804899     32 YBR298C
        # 866: 4497 chrII 809353 809377  -  809377 1.788977   1.490814 809353 809377     25 YBR300C
        # 867: 4498 chrII 809707 809748  -  809707 3.876117   2.087139 809707 809724     18 YBR300C
        # 868: 4503 chrII 811446 811512  -  811486 607.059367  313.965377 811485 811494     10 YBR302C
        # 
        # $treat
        # cluster chr start end strand dominant_tss  tags tags.dominant_tss q_0.1 q_0.9 interquantile_width  gene
        # 1:  1 chrI 1828 1841  +   1828 1.078417   0.359473 1828 1841     14 YAL067W-A
        # 2:  6 chrI 9327 9476  +   9359 10.604449   1.617628 9327 9452     126 YAL066W
        # 3:  8 chrI 11259 11343  +  11329 54.639873   20.130481 11288 11329     42 YAL064W-B
        # 4:  9 chrI 11651 11762  +  11651 1.977099   0.359473 11651 11750     100 YAL064W-B
        # 5:  10 chrI 11861 11934  +  11919 2.875780   0.898682 11889 11923     35 YAL064W-B
        # ---                                 
        # 816: 4488 chrII 801017 801061  -  801023 6.290772   1.797364 801022 801058     37 YBR296C-A
        # 817: 4494 chrII 804851 804952  -  804895 144.867571   50.326204 804872 804901     30 YBR298C
        # 818: 4497 chrII 809377 809389  -  809377 1.078419   0.718946 809377 809389     13 YBR300C
        # 819: 4498 chrII 809707 809710  -  809707 1.258155   0.898682 809707 809710     4 YBR300C
        # 820: 4503 chrII 811472 811503  -  811486 449.700575  234.196583 811486 811494     9 YBR302CThe 

  Instead of visualizing TSSs and core promoters in the UCSC Genome Browser or IGV, plotTSS function is able to generate publish ready figures when list of interested genes are provided and plotting region is specified.
                  
  A function of exportClustersTable is provided to export cluster tables with associated gene information.
        
        exportClustersTable(myTSSr, data = "assigned")

* Differential expression analysis

  The number of tags at each TSS reflects the number of transcripts initiated at the TSS. Thus, TSS data can be used for expression profiling. With specified sample pairs for comparison, deGene function counts raw tags of each consensus clusters and utilizes the DESeq2 package (Love, Huber et al. 2014) for differential expression analysis. 
        
        deGene(myTSSr,comparePairs=list(c("control","treat")), pval = 0.01,useMultiCore=FALSE, numCores=NULL)
        
  Differential expression analysis results can be visualized by plotDE function which generates a volcano plots. Names of genes differential expressed between the compared pairs are displayed on the dots when the withGeneName argument is set as TRUE.  

        plotDE(myTSSr, withGeneName = "TRUE",xlim=c(-2.5, 2.5),ylim=c(0,10))


  Differential expression analysis results can also be exported to a text file with exportDETable function. 

        exportDETable(myTSSr, data = "sig")

![05_Volcano_plot](https://github.com/Linlab-slu/TSSr/raw/master/vignettes/figures/05_Volcano_plot.png){width=50%}

* Core promoter shifts


NEWS
0.1.0 (03-08-2021)
+ Code has been tested and is stable. Preparing for submission to Bioconductor once vignette (and various other requirements) is complete.


R/assign2gene.R
#####################################################################################################
#####################################################################################################
##.assign2gene function assign clusters to reference
##.assign2gene function takes two input files, clusters and refernce
##users need to provide organism scientific name like Saccharomyces cerevisiae
##reference could be "gff3" or "gtf" file, or an URL
##clusters table has 11 columns (cluster,chr,start,end,strand,dominant_tss,tpm,tpm.dominant_tss,q_0.1,q_0.9,interquantile_width) 
##run script with the following example command:
##.assign2gene(clusters,reference,organim,upstream=1000, upstreamOverlap = 500)


.assign2gene <- function(clusters,reference,organim,upstream=1000, upstreamOverlap = 500){
  setDT(clusters)
  clusters[, "subset" := paste0(chr,"_",strand)]
  txdb <- suppressWarnings(makeTxDbFromGFF(reference, organism, format = "auto"))
  ref <- setDT(as.data.frame(genes(txdb)))
  ref[, "subset" := paste0(seqnames,"_",strand)]
  setkey(ref,subset)
  setkey(clusters,subset)
  asn <- lapply(as.list(clusters[,unique(subset)]), function(x) {
    cs <- clusters[.(x)]
    cs[,subset:= NULL]
    colnames(cs)[3:4] <- c("start.c","end.c")
    gr <- makeGRangesFromDataFrame(cs, keep.extra.columns= T, start.field = "dominant_tss", end.field = "dominant_tss")
    ref_sub <- subset(ref[ref$subset == x,]) ## why ref_sub <- ref[.(x)] not working when x == "chrM_-"?
     if(nrow(ref_sub) == 0){
      mcols(gr)[,"gene"] <- NA
    }else{
      ref_sub$start <- as.numeric(ref_sub$start)
      setorder(ref_sub,start)
      if(cs$strand[1] == "+"){
        setorder(ref_sub,start)
        ref_sub[,end.b:=data.table::shift(end, 1, fill = 0,type='lag')]
        ref_sub[,width:=data.table::shift(width, 1, fill = 1000,type='lag')]
        ref_sub[,dis := start - end.b]
        ref_sub[,dis:= ifelse(dis < 0, 0,dis)]
        ref_sub[,up:= ifelse(dis > upstream, upstream, 
                             ifelse(dis + width  <= upstreamOverlap, dis + width -1,
                                    ifelse(dis < upstreamOverlap, upstreamOverlap, dis)))]
        ref_sub[,end:= start] ##start - 1 -> start April10
        ref_sub[,start := end - up +1]
      }else{
        setorder(ref_sub,end)
        ref_sub[,end.b:=data.table::shift(start, 1, fill = 0,type='lead')]
        ref_sub[(.N),end.b := end + 1000]
        ref_sub[,width:=data.table::shift(width, 1, fill = 1000,type='lead')]
        ref_sub[,dis := end.b - end] ##start -> end April10
        ref_sub[,dis:= ifelse(dis < 0,0,dis)]
        ref_sub[,up:= ifelse(dis > upstream, upstream, 
                             ifelse(dis + width  <= upstreamOverlap, dis + width -1,
                                    ifelse(dis < upstreamOverlap, upstreamOverlap, dis)))]
        ref_sub[,start:= end]##end + 1 -> end April10
        ref_sub[,end:= start + up -1]
      }
      rownames(ref_sub) <- ref_sub$gene_id
      ref_sub <- makeGRangesFromDataFrame(ref_sub, keep.extra.columns= F)
      hits <- findOverlaps(gr,ref_sub)
      hits <- breakTies(hits, method = "first")
      hits <- methods::as(hits, "List")
      hits <- extractList(names(ref_sub), hits)
      hits <- as.character(hits)
      mcols(gr)[,"gene"] <- hits
    }
    gr <- as.data.frame(gr)
    gr$dominant_tss <- gr$start
    colnames(gr)[c(1,7,8)] <- c("chr","start","end")
    gr <- gr[,c(6,1,7,8,5,15,9:14)]
    return(gr)
  })
  setorder(do.call("rbind",asn), cluster)
}



#####################################################################################################

